File: pcm2pwle/frequency_amplitude_extraction.py
#  Copyright 2026 Google LLC
#
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#
#      https://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.

"""Extract frequency and amplitude envelope, and control points from the vibration data."""

import copy
import logging

import numpy as np
import pandas as pd
from scipy import signal

from pcm2pwle import common_utils


def vib_data_preprocess(
    data: np.ndarray,
    rate: int,
    preprocess_lowpass_cutoff: int = 500,
    crop_s: float | None = None,
) -> np.ndarray:
  """Preprocess the vibration data.

  Args:
    data: The vibration data.
    rate: The sampling rate of the data.
    preprocess_lowpass_cutoff: The lowpass cutoff frequency for the data.
    crop_s: The duration seconds to keep from the start of the signal. Default
      is None (no cropping). Sometimes the signals can be very long to do the
      conversion. This allows users to cut the signal before analysis.

  Returns:
    The preprocessed vibration data.
  """
  data = copy.deepcopy(data)

  # Remove high freq noise
  if preprocess_lowpass_cutoff is not None:
    b, a = signal.butter(5, preprocess_lowpass_cutoff / (rate / 2), btype='low')
    data = signal.filtfilt(b, a, data)

  # Crop the data to keep only the first crop_s seconds
  if crop_s is not None:
    end_idx = np.min((len(data), int(crop_s * rate)))
    data = data[:end_idx]

  # Normalize data to [-1, 1]
  max_abs_data = np.max(np.abs(data))
  if max_abs_data > 0:
    data = data / max_abs_data

  return data


def _calculate_freq_env(
    data: np.ndarray,
    rate: int,
    envelope_lowpass_cutoff: float = 100
):
  """Calculates the frequency envelope of the vibration data."""
  analytic_signal = signal.hilbert(data)
  amp_envelope_raw = np.abs(analytic_signal)
  ins_phase = np.unwrap(np.angle(analytic_signal))
  freq_envelope_raw = np.diff(ins_phase) * rate / (2.0 * np.pi)
  freq_envelope_raw = np.insert(freq_envelope_raw, 0, freq_envelope_raw[0])

  freq_envelope_raw[freq_envelope_raw < 0] = np.nan
  upper_bound = 500

  logging.info('rm_outliers: kept frequency range: 0 %s', upper_bound)
  freq_envelope_raw[freq_envelope_raw > upper_bound] = np.nan
  freq_envelope_raw = pd.Series(freq_envelope_raw).interpolate().values
  freq_envelope_raw[np.isnan(freq_envelope_raw)] = 0

  sos = signal.butter(
      5, envelope_lowpass_cutoff / (rate / 2), btype='low', output='sos'
  )
  amp_envelope = signal.sosfiltfilt(sos, amp_envelope_raw)
  freq_envelope = signal.sosfiltfilt(sos, freq_envelope_raw)

  return amp_envelope, freq_envelope


def extract_vib_amp_freq(
    data: np.ndarray, rate: int, envelope_lowpass_cutoff: float = 100
) -> tuple[np.ndarray, np.ndarray]:
  """Extracts the vibration amplitude and frequency envelope from the data.

  Args:
    data: The vibration data.
    rate: The sampling rate of the data.
    envelope_lowpass_cutoff: The lowpass cutoff frequency for the envelope.

  Returns:
    A tuple of vibration amplitude and frequency envelopes.
  """
  data = copy.deepcopy(data)
  amp_envelope, freq_envelope = _calculate_freq_env(
      data, rate, envelope_lowpass_cutoff=envelope_lowpass_cutoff
  )

  chunks = common_utils.zero_out_low_amplitude_chunks(
      data, rate, amplitude_threshold_fraction=0.01, duration_threshold_ms=20
  )
  for _, (start, end) in enumerate(chunks):
    amp_envelope[start:end] = 0.0
    freq_envelope[start:end] = 0.0

  amp_envelope[-1] = 0.0
  freq_envelope[-1] = 0.0

  return amp_envelope, freq_envelope

File: pcm2pwle/frequency_amplitude_extraction.cc
#include "pcm2pwle/frequency_amplitude_extraction.h"

#include <algorithm>
#include <cmath>
#include <cstddef>
#include <optional>
#include <utility>
#include <vector>

#include "pcm2pwle/common_utils.h"
#include "pcm2pwle/dsp_utils.h"

namespace pcm2pwle {

std::vector<double> VibDataPreprocess(const std::vector<double>& data, int rate,
                                      double preprocessLowpassCutoff,
                                      std::optional<double> cropS) {
  if (data.empty()) return {};
  std::vector<double> working = data;

  if (preprocessLowpassCutoff > 0.0) {
    auto sos = ButterworthLowpassSos(5, preprocessLowpassCutoff, rate);
    working = Sosfiltfilt(sos, working);
  }

  if (cropS.has_value() && cropS.value() > 0.0) {
    size_t endIdx =
        std::min(working.size(), static_cast<size_t>(cropS.value() * rate));
    working.resize(endIdx);
  }

  double maxAbs = 0.0;
  for (double val : working) {
    double absVal = std::abs(val);
    if (absVal > maxAbs) {
      maxAbs = absVal;
    }
  }

  if (maxAbs > 0.0) {
    for (double& val : working) {
      val /= maxAbs;
    }
  }

  return working;
}

static std::pair<std::vector<double>, std::vector<double>> CalculateFreqEnv(
    const std::vector<double>& data, int rate, double envelopeLowpassCutoff) {
  int n = data.size();
  if (n == 0) return {{}, {}};

  std::vector<double> real, imag;
  Hilbert(data, &real, &imag);

  std::vector<double> ampEnvelopeRaw(n);
  std::vector<double> insPhase(n);
  for (size_t i = 0; i < data.size(); ++i) {
    ampEnvelopeRaw[i] = std::sqrt(real[i] * real[i] + imag[i] * imag[i]);
    insPhase[i] = std::atan2(imag[i], real[i]);
  }

  std::vector<double> unwrappedPhase = Unwrap(insPhase);
  std::vector<double> freqEnvelopeRaw(n);
  for (size_t i = 1; i < data.size(); ++i) {
    freqEnvelopeRaw[i] =
        ((unwrappedPhase[i] - unwrappedPhase[i - 1]) * rate) / (2.0 * M_PI);
  }
  freqEnvelopeRaw[0] = n > 1 ? freqEnvelopeRaw[1] : 0.0;

  double upperBound = 500.0;
  std::vector<int> validIndices;
  std::vector<double> validFreqs;
  for (size_t i = 0; i < data.size(); ++i) {
    double f = freqEnvelopeRaw[i];
    if (f >= 0.0 && f <= upperBound && !std::isnan(f) && std::isfinite(f)) {
      validIndices.push_back(i);
      validFreqs.push_back(f);
    }
  }

  if (!validIndices.empty() && validIndices.size() < static_cast<size_t>(n)) {
    size_t validIdx = 0;
    for (size_t i = 0; i < data.size(); ++i) {
      if (i <= validIndices.front()) {
        freqEnvelopeRaw[i] = validFreqs.front();
      } else if (i >= validIndices.back()) {
        freqEnvelopeRaw[i] = validFreqs.back();
      } else {
        while (validIdx + 1 < validIndices.size() &&
               validIndices[validIdx + 1] <= i) {
          validIdx++;
        }
        double x0 = validIndices[validIdx];
        double x1 = validIndices[validIdx + 1];
        double y0 = validFreqs[validIdx];
        double y1 = validFreqs[validIdx + 1];
        double t = (i - x0) / ((x1 - x0) != 0.0 ? (x1 - x0) : 1.0);
        freqEnvelopeRaw[i] = y0 + t * (y1 - y0);
      }
    }
  } else if (validIndices.empty()) {
    std::fill(freqEnvelopeRaw.begin(), freqEnvelopeRaw.end(), 0.0);
  }

  auto sos = ButterworthLowpassSos(5, envelopeLowpassCutoff, rate);
  std::vector<double> ampEnvelope = Sosfiltfilt(sos, ampEnvelopeRaw);
  std::vector<double> freqEnvelope = Sosfiltfilt(sos, freqEnvelopeRaw);

  return {ampEnvelope, freqEnvelope};
}

std::pair<std::vector<double>, std::vector<double>> ExtractVibAmpFreq(
    const std::vector<double>& data, int rate, double envelopeLowpassCutoff) {
  auto [ampEnvelope, freqEnvelope] =
      CalculateFreqEnv(data, rate, envelopeLowpassCutoff);

  auto chunks = ZeroOutLowAmplitudeChunks(data, rate, 0.01, 20.0);
  for (const auto& chunk : chunks) {
    for (int i = chunk.first;
         i < chunk.second && i < static_cast<int>(ampEnvelope.size()); ++i) {
      ampEnvelope[i] = 0.0;
      freqEnvelope[i] = 0.0;
    }
  }

  if (!ampEnvelope.empty()) {
    ampEnvelope.back() = 0.0;
    freqEnvelope.back() = 0.0;
  }

  return {ampEnvelope, freqEnvelope};
}

}  // namespace pcm2pwle
